src/utils/saml.ts

```diff
@@ -1118,395 +81,542 @@ export const getPreValidateResponse = (
+      TE.chain(__ =>
+        pipe(
+          TE.fromEither(
+            pipe(
+              UTCISODateFromString.decode(
+                __.RequestAuthnRequest.getAttribute("IssueInstant")
+              ),
+              E.mapLeft(
+                () =>
+                  new Error(
+                    "IssueInstant into the Request must be a valid UTC string"
+                  )
+              )
             )
-          ).map(RequestIssueInstant => ({ ...__, RequestIssueInstant }))
+          ),
+          TE.map(RequestIssueInstant => ({ ...__, RequestIssueInstant }))
         )
-    )
-    .chain(_ =>
-      fromEitherToTaskEither(
-        fromPredicate<Error, Date>(
-          _1 => _1.getTime() <= _.IssueInstant.getTime(),
-          () =>
-            new Error("Request IssueInstant must after Request IssueInstant")
-        )(_.RequestIssueInstant)
-      ).map(() => _)
-    )
-    .chain(_ =>
-      fromEitherToTaskEither(
-        fromPredicate<Error, Date>(
-          _1 => _1.getTime() <= _.AssertionIssueInstant.getTime(),
+      )
+    );
+
+  const stepFive = (_: IStepFour): TaskEither<Error, IStepFive> =>
+    pipe(
+      TE.fromEither(
+        pipe(
+          _.RequestIssueInstant,
+          E.fromPredicate(
+            _1 => _1.getTime() <= _.IssueInstant.getTime(),
+            () =>
+              new Error("Request IssueInstant must after Request IssueInstant")
+          )
+        )
+      ),
+      TE.map(() => _)
+    );
+
+  const stepSix = (_: IStepFive): TaskEither<Error, IStepSix> =>
+    pipe(
+      TE.fromEither(
+        pipe(
+          _.RequestIssueInstant,
+          E.fromPredicate(
+            _1 => _1.getTime() <= _.AssertionIssueInstant.getTime(),
+            () =>
+              new Error(
+                "Assertion IssueInstant must after Request IssueInstant"
+              )
+          )
+        )
+      ),
+      TE.map(() => _)
+    );
+
+  const stepSeven = (_: IStepSix): TaskEither<Error, IStepSeven> =>
+    TE.fromEither(
+      pipe(
+        E.fromOption(
           () =>
-            new Error("Assertion IssueInstant must after Request IssueInstant")
-        )(_.RequestIssueInstant)
-      ).map(() => _)
-    )
-    .chain(_ =>
-      fromEitherToTaskEither(
-        fromOption(
-          new Error("Missing AuthnContextClassRef inside cached SAML Response")
+            new Error(
+              "Missing AuthnContextClassRef inside cached SAML Response"
+            )
         )(
-          fromNullable(
+          O.fromNullable(
             _.RequestAuthnRequest.getElementsByTagNameNS(
               SAML_NAMESPACE.ASSERTION,
               "AuthnContextClassRef"
             ).item(0)
           )
-        )
-          .chain(
-            fromPredicate(
-              not(isEmptyNode),
-              () => new Error("Subject element must be not empty")
-            )
+        ),
+        E.chain(
+          E.fromPredicate(
+            not(isEmptyNode),
```

Comment:
`not` function is deprecated, use `not` from `fp-ts/lib/Predicate` module
```suggestion
            predicate.not(isEmptyNode),
```

Reply:
fixed, thank you!